packages/std/src/init_handle.rs

```diff
@@ -10,12 +11,15 @@ use crate::types::HumanAddr;
 
 #[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
 #[serde(rename_all = "snake_case")]
-pub enum CosmosMsg {
+// See https://github.com/serde-rs/serde/issues/1296 why we cannot add trait bounds to T
+pub enum CosmosMsg<T = NativeMsg>
+where
+    T: Clone + fmt::Debug + PartialEq + JsonSchema,
+{
```

Comment:
So this derives `Serialize` for the custom `T`. When the encoded message is passed untouched through go-cosmwasm, there needs to be a Go implementation of `Deserialize` that exactly matches what serde-json-wasm does. I think this is pretty hard, given the encoding is hidden in some external library.

Reply:
This also forces the usage of JSON for custom message encoding.

As a system designer, I'd rather implement

```rust
CosmosMsg::Native(NativeMsg::Raw(encode_my_custom_message(msg_object)))
```

in the contract and write a reverse operation in Go instead of depending in json-serde-wasm, which might be replaced at any time.

Reply:
>  So this derives Serialize for the custom T. When the encoded message is passed untouched through go-cosmwasm, there needs to be a Go implementation of Deserialize that exactly matches what serde-json-wasm does. I think this is pretty hard, given the encoding is hidden in some external library.

This doesn't derive Serialize, but requires that T derives Serialize explicitly (there is some issue about why I cannot specify that as a trait bound, but the compiler does enforce it)

Reply:
> As a system designer, I'd rather implement

> `CosmosMsg::Native(NativeMsg::Raw(encode_my_custom_message(msg_object)))`

> in the contract and write a reverse operation in Go instead of depending in json-serde-wasm, which might be replaced at any time.

IMO that is very nasty UX. I have done some with the cosmos sdk message for reflect and it is a bit of a pain. (and that was outside of the contract). I would leave this space open for any custom encoding, but not as the only way.

Reply:
Maybe I should explain the motivation better:

My idea is that we add some standard top-level types like `BankMsg` and soon `StakingMsg`, that provides a very nice way of interacting with those standard modules. We work on keeping that easy to use and maybe polish a bit more. But we have to be conservative on what we add, as we expect it to be immutable and supported the same way by (almost) all chains.

However, a chain will have some custom modules, and would like to add their own types to `CosmosMsg`, like `SwapMsg` or `VoteMsg`. They would like a similar DevUX for contract builders like using `BankMsg` or `StakingMsg`. But we cannot encode all this in `cosmwasm_std`.

My idea here was to make `Native` extensible to allow such chain-specific APIs to be developed. Then they expose a crate, eg. `cw-terra`, which exposes `TerraMsg` and when you use `CosmosMsg<TerraMsg>`, you can use all those feautures, so easy for contract builders.

I would keep `NativeMsg::Raw` as well (most chains should include that along with custom APIs), if you want low-level access, but there is no guarantees there. A theoretical `CosmosMsg::Native(TerraMsg)` type could easily offer stability / immutability guarantees *on the terra chain*

Reply:
> This doesn't derive Serialize, but requires that T derives Serialize explicitly (there is some issue about why I cannot specify that as a trait bound, but the compiler does enforce it)

okay, but

```rust
let v = to_vec(&res).unwrap();
```

in `do_handle` enforces encoding via (currently) serde_json_wasm.

Reply:
Okay, you mean on the contract. Yes, we do require JSON like all the other types we use here. I see no problem with that now.

`NativeMsg::Raw(Binary)` gives the option to use any other encoding if you wish to do a separate encoding step.

The extensible `CosmosMsg::Native` variant was conceived to work like `BankMsg` or `HandleMsg` or any other message we typically send. And then require custom code in the go module to translate it into a raw native type.

Maybe it is the term `Native` you are concerned with? For me this means calling into a  *native module*, not using *native encoding*.

Reply:
> Yes, we do require JSON like all the other types we use here. I see no problem with that now.

We not only require JSON. There are a million ways to encode a Rust struct to JSON, and we require the one provided by the [broken and incomplete](https://github.com/CosmWasm/cosmwasm/issues/110) json-serde-wasm imlementation. Also there are no specs for this encoding. And we do so in a hidden, implicit way without any warning.

While this is fine for things controlled by us, I don't think we should extend the usage of this encoder to (1) other developers, (2) other languages, (3) unknown use cases/structures.

> Maybe it is the term Native you are concerned with?

The term is fine for now. It is more about the required reverse engineering required to write a decoder as well as the limitations of the encoder.

Reply:
If this is the main blocker, I would do some pending cleanup (mentioned above) and generate schema types for all custom messages. This is no different / broken encoding than `CosmosMsg` is right now, and using `Raw(Binary)` as a variant is always a way to embed a different encoding.

There was no big issues using the json schema files in JS/TS land, so I don't see this as worse. I do see a good point of tackling #110 though.

Reply:
I'll look into #110.

The JSON schema is indeed a very good way to ensure consistency between the encoder and the decoder (instead of just brute force a decoder from example data)